// baseui/codemods/shared.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.containsFlowComment = containsFlowComment;
exports.getStyledImportName = getStyledImportName;

/*
Copyright (c) 2018-2020 Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
function containsFlowComment(j, root) {
  var getBodyNode = function getBodyNode() {
    return root.find(j.Program).get('body', 0).node;
  };

  var comments = getBodyNode().comments || [];
  return comments.filter(function (e) {
    return e.value.indexOf('@flow') !== -1;
  }).length > 0;
}

function getStyledImportName(j, root) {
  var baseImports = root.find(j.ImportDeclaration, {
    source: {
      value: 'baseui'
    }
  });

  if (baseImports.length) {
    var styledImports = baseImports.find(j.ImportSpecifier, {
      imported: {
        name: 'styled'
      }
    });

    if (styledImports.length) {
      return styledImports.get(0).node.local.name;
    }
  }

  return null;
}
