// baseui/esm/data-table/cell-shell.js
function _objectSpread2(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

/*
Copyright (c) 2018-2020 Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import * as React from 'react';
import { Checkbox } from '../checkbox/index.js';
import { useStyletron } from '../styles/index.js';
var CellShell = React.forwardRef(function (props, ref) {
  var _useStyletron = useStyletron(),
      _useStyletron2 = _slicedToArray(_useStyletron, 2),
      css = _useStyletron2[0],
      theme = _useStyletron2[1];

  return React.createElement("div", {
    ref: ref,
    className: css(_objectSpread2({}, theme.typography.font100, {
      boxSizing: 'border-box',
      display: props.isMeasured ? 'inline-block' : null,
      paddingTop: theme.sizing.scale300,
      paddingLeft: theme.sizing.scale500,
      paddingBottom: theme.sizing.scale300,
      paddingRight: theme.sizing.scale500,
      width: props.isMeasured ? null : '100%'
    }))
  }, React.createElement("div", {
    className: css({
      display: 'flex',
      alignItems: 'center'
    })
  }, Boolean(props.onSelect) && React.createElement(Checkbox, {
    onChange: props.onSelect,
    checked: props.isSelected,
    overrides: {
      Checkmark: {
        style: {
          marginTop: null,
          marginBottom: null
        }
      }
    }
  }), props.children));
});
CellShell.displayName = 'CellShell';
export default CellShell;
