// baseui/esm/data-table/header-cell.js
function _objectSpread2(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

/*
Copyright (c) 2018-2020 Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import * as React from 'react';
import { Checkbox } from '../checkbox/index.js';
import { useStyletron } from '../styles/index.js';
import ChevronDown from '../icon/chevron-down.js';
import ChevronUp from '../icon/chevron-up.js';
import { SORT_DIRECTIONS } from './constants.js';
var HeaderCell = React.forwardRef(function (props, ref) {
  var _useStyletron = useStyletron(),
      _useStyletron2 = _slicedToArray(_useStyletron, 2),
      css = _useStyletron2[0],
      theme = _useStyletron2[1];

  var backgroundColor = props.isHovered ? theme.colors.mono200 : theme.colors.mono100;
  return React.createElement("div", {
    ref: ref,
    role: "button",
    tabIndex: "0",
    className: css(_objectSpread2({}, theme.typography.font150, {
      alignItems: 'center',
      backgroundColor: backgroundColor,
      boxSizing: 'border-box',
      cursor: props.sortable ? 'pointer' : null,
      display: props.isMeasured ? 'inline-flex' : 'flex',
      flexGrow: 1,
      height: '100%',
      paddingLeft: theme.sizing.scale500,
      paddingRight: theme.sizing.scale500,
      flexWrap: 'nowrap',
      whiteSpace: 'nowrap'
    })),
    onMouseEnter: props.onMouseEnter,
    onMouseLeave: props.onMouseLeave,
    onKeyUp: function onKeyUp(event) {
      if (event.key === 'Enter') {
        props.onSort(props.index);
      }
    },
    onClick: function onClick() {
      return props.onSort(props.index);
    }
  }, props.isSelectable && React.createElement(Checkbox, {
    onChange: function onChange(e) {
      if (props.isSelectedAll || props.isSelectedIndeterminate) {
        props.onSelectNone();
      } else {
        props.onSelectAll();
      }
    },
    checked: props.isSelectedAll || props.isSelectedIndeterminate,
    isIndeterminate: props.isSelectedIndeterminate
  }), props.title, React.createElement("div", {
    className: css({
      position: 'relative',
      width: '100%',
      display: 'flex',
      alignItems: 'center'
    })
  }, (props.isHovered || props.sortDirection) && props.sortable && React.createElement("div", {
    style: {
      backgroundColor: backgroundColor,
      display: 'flex',
      alignItems: 'center',
      position: 'absolute',
      right: -4
    }
  }, (props.sortDirection === SORT_DIRECTIONS.DESC || !props.sortDirection) && React.createElement(ChevronDown, {
    color: props.sortDirection ? theme.colors.primary : theme.colors.mono600
  }), props.sortDirection === SORT_DIRECTIONS.ASC && React.createElement(ChevronUp, {
    color: props.sortDirection ? theme.colors.primary : theme.colors.mono600
  }))));
});
HeaderCell.displayName = 'HeaderCell';
export default HeaderCell;
