// baseui/es/tree-view/tree-node.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/*
Copyright (c) 2018-2020 Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import * as React from 'react';
import { StyledTreeItemList, StyledTreeItem } from './styled-components.js';
import StyledTreeLabel from './tree-label.js';
import { getOverride, getOverrideProps } from '../helpers/overrides.js';
export default class TreeNode extends React.Component {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "onToggle", () => {
      const {
        onToggle,
        node
      } = this.props;

      if (onToggle) {
        onToggle(node);
      }
    });
  }

  render() {
    const {
      node,
      onToggle,
      overrides = {},
      renderAll
    } = this.props;
    const {
      children,
      isExpanded,
      label
    } = node;
    const hasChildren = children && children.length !== 0;
    const {
      TreeItemList: TreeItemListOverride,
      TreeItem: TreeItemOverride,
      TreeLabel: TreeLabelOverride
    } = overrides;
    const TreeItemList = getOverride(TreeItemListOverride) || StyledTreeItemList;
    const TreeItem = getOverride(TreeItemOverride) || StyledTreeItem;
    const TreeLabel = getOverride(TreeLabelOverride) || StyledTreeLabel;
    return React.createElement(TreeItem, _extends({
      role: "treeitem",
      "aria-expanded": isExpanded ? true : false,
      $isLeafNode: !hasChildren
    }, getOverrideProps(TreeItemOverride)), React.createElement(TreeLabel, _extends({
      onClick: this.onToggle,
      node: node,
      hasChildren: hasChildren,
      isExpanded: isExpanded,
      label: label,
      overrides: overrides
    }, getOverrideProps(TreeLabelOverride))), children && (isExpanded || renderAll) && React.createElement(TreeItemList, _extends({
      role: "group",
      $isChildNode: true,
      $expanded: !!isExpanded
    }, getOverrideProps(TreeItemListOverride)), children.map((node, index) => React.createElement(TreeNode, {
      renderAll: renderAll,
      key: index,
      node: node,
      onToggle: onToggle,
      overrides: overrides
    }))));
  }

}
