// baseui/esm/link/styled-components.js
function _objectSpread2(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/*
Copyright (c) 2018-2020 Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import { styled } from '../styles/index.js';
export var Link = styled('a', function (_ref) {
  var $theme = _ref.$theme;
  var colors = $theme.colors,
      typography = $theme.typography,
      animation = $theme.animation;
  return _objectSpread2({
    color: colors.linkText
  }, typography.font350, {
    fontSize: 'inherit',
    lineHeight: 'inherit',
    transitionProperty: 'color',
    transitionDuration: animation.timing100,
    transitionTimingFunction: animation.easeOutCurve,
    ':visited': {
      color: colors.linkVisited
    },
    ':hover': {
      color: colors.linkHover
    },
    ':active': {
      color: colors.linkActive
    }
  });
});
Link.displayName = "Link";
