// baseui/esm/themes/light-theme/light-theme.js
function _objectSpread2(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/*
Copyright (c) 2018-2020 Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import colors from './color-tokens.js';
import getSemanticColors from './color-semantic-tokens.js';
import getColorComponentTokens from './color-component-tokens.js';
import getDeprecatedSemanticColors from './color-deprecated-semantic-tokens.js';
import borders from '../shared/borders.js';
import lighting from './lighting.js';
import getTypography from '../shared/typography.js';
import animation from '../shared/animation.js';
import breakpoints from '../shared/breakpoints.js';
import grid from '../shared/grid.js';
import mediaQuery from '../shared/media-query.js';
import sizing from '../shared/sizing.js';
export var LightTheme = {
  name: 'light-theme',
  colors: _objectSpread2({}, colors, {}, getColorComponentTokens(), {}, getDeprecatedSemanticColors(), {}, getSemanticColors()),
  animation: animation,
  breakpoints: breakpoints,
  borders: borders,
  direction: 'auto',
  grid: grid,
  lighting: lighting,
  mediaQuery: mediaQuery,
  sizing: sizing,
  typography: getTypography(),
  // TODO(#2318) Remove it in the next v10 major version.
  // Do not use.
  zIndex: {
    modal: 2000
  }
};
