// baseui/card/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Card", {
  enumerable: true,
  get: function get() {
    return _card.default;
  }
});
Object.defineProperty(exports, "hasThumbnail", {
  enumerable: true,
  get: function get() {
    return _card.hasThumbnail;
  }
});
Object.defineProperty(exports, "StyledAction", {
  enumerable: true,
  get: function get() {
    return _styledComponents.Action;
  }
});
Object.defineProperty(exports, "StyledBody", {
  enumerable: true,
  get: function get() {
    return _styledComponents.Body;
  }
});
Object.defineProperty(exports, "StyledContents", {
  enumerable: true,
  get: function get() {
    return _styledComponents.Contents;
  }
});
Object.defineProperty(exports, "StyledHeaderImage", {
  enumerable: true,
  get: function get() {
    return _styledComponents.HeaderImage;
  }
});
Object.defineProperty(exports, "StyledThumbnail", {
  enumerable: true,
  get: function get() {
    return _styledComponents.Thumbnail;
  }
});
Object.defineProperty(exports, "StyledTitle", {
  enumerable: true,
  get: function get() {
    return _styledComponents.Title;
  }
});
Object.defineProperty(exports, "StyledRoot", {
  enumerable: true,
  get: function get() {
    return _styledComponents.Root;
  }
});
Object.defineProperty(exports, "StyledWrapper", {
  enumerable: true,
  get: function get() {
    return _styledComponents.Root;
  }
});

var _card = _interopRequireWildcard(require("./card.js"));

var _styledComponents = require("./styled-components.js");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }
