// baseui/card/styled-components.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Title = exports.Thumbnail = exports.Root = exports.HeaderImage = exports.Contents = exports.Body = exports.Action = void 0;

var _index = require("../styles/index.js");

function _objectSpread2(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Action = (0, _index.styled)('div', function (_ref) {
  var $theme = _ref.$theme;
  return _objectSpread2({}, $theme.typography.LabelMedium);
});
exports.Action = Action;
Action.displayName = "Action";
var Body = (0, _index.styled)('div', function (_ref2) {
  var $theme = _ref2.$theme;
  return _objectSpread2({
    marginBottom: $theme.sizing.scale600,
    color: $theme.colors.contentPrimary
  }, $theme.typography.ParagraphMedium);
});
exports.Body = Body;
Body.displayName = "Body";
var Contents = (0, _index.styled)('div', function (_ref3) {
  var $theme = _ref3.$theme;
  return {
    marginLeft: $theme.sizing.scale600,
    marginTop: $theme.sizing.scale600,
    marginRight: $theme.sizing.scale600,
    marginBottom: $theme.sizing.scale600
  };
});
exports.Contents = Contents;
Contents.displayName = "Contents";
var HeaderImage = (0, _index.styled)('img', function (_ref4) {
  var $theme = _ref4.$theme;
  return {
    borderTopLeftRadius: $theme.borders.surfaceBorderRadius,
    borderTopRightRadius: $theme.borders.surfaceBorderRadius,
    objectFit: 'contain',
    maxWidth: '100%'
  };
});
exports.HeaderImage = HeaderImage;
HeaderImage.displayName = "HeaderImage";
var Root = (0, _index.styled)('div', function (_ref5) {
  var $theme = _ref5.$theme;
  return {
    borderLeftWidth: '2px',
    borderTopWidth: '2px',
    borderRightWidth: '2px',
    borderBottomWidth: '2px',
    borderLeftStyle: 'solid',
    borderTopStyle: 'solid',
    borderRightStyle: 'solid',
    borderBottomStyle: 'solid',
    borderColor: $theme.colors.borderOpaque,
    borderTopLeftRadius: $theme.borders.surfaceBorderRadius,
    borderTopRightRadius: $theme.borders.surfaceBorderRadius,
    borderBottomLeftRadius: $theme.borders.surfaceBorderRadius,
    borderBottomRightRadius: $theme.borders.surfaceBorderRadius,
    backgroundColor: $theme.colors.backgroundPrimary
  };
});
exports.Root = Root;
Root.displayName = "Root";
var Thumbnail = (0, _index.styled)('img', function (_ref6) {
  var $theme = _ref6.$theme;
  return _objectSpread2({
    float: 'right',
    height: $theme.sizing.scale2400,
    width: $theme.sizing.scale2400,
    objectFit: 'cover',
    borderTopLeftRadius: $theme.borders.surfaceBorderRadius,
    borderTopRightRadius: $theme.borders.surfaceBorderRadius,
    borderBottomLeftRadius: $theme.borders.surfaceBorderRadius,
    borderBottomRightRadius: $theme.borders.surfaceBorderRadius
  }, $theme.borders.border200, {
    margin: "0 0 ".concat($theme.sizing.scale500, " ").concat($theme.sizing.scale500)
  });
});
exports.Thumbnail = Thumbnail;
Thumbnail.displayName = "Thumbnail";
var Title = (0, _index.styled)('h1', function (_ref7) {
  var $theme = _ref7.$theme;
  return _objectSpread2({}, $theme.typography.HeadingSmall, {
    color: $theme.colors.contentPrimary,
    fontWeight: 500,
    marginLeft: 0,
    marginTop: 0,
    marginRight: 0,
    marginBottom: $theme.sizing.scale300,
    paddingLeft: 0,
    paddingTop: 0,
    paddingRight: 0,
    paddingBottom: 0
  });
});
exports.Title = Title;
Title.displayName = "Title";
