// baseui/esm/layout-grid/grid.js
/*
Copyright (c) 2018-2020 Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import React from 'react';
import { StyledGrid, StyledCell } from './styled-components.js';
export default function Grid(_ref) {
  var align = _ref.align,
      behavior = _ref.behavior,
      children = _ref.children,
      gridColumns = _ref.gridColumns,
      gridGaps = _ref.gridGaps,
      gridGutters = _ref.gridGutters,
      gridMargins = _ref.gridMargins,
      gridMaxWidth = _ref.gridMaxWidth,
      gridUnit = _ref.gridUnit;
  return React.createElement(StyledGrid, {
    $behavior: behavior,
    $gridMargins: gridMargins,
    $gridGutters: gridGutters,
    $gridMaxWidth: gridMaxWidth,
    $gridUnit: gridUnit,
    $align: align
  }, React.Children.map(children, function (child) {
    return React.createElement(StyledCell, {
      $align: child.props.align,
      $order: child.props.order,
      $skip: child.props.skip,
      $span: child.props.span,
      $gridColumns: gridColumns,
      $gridGutters: gridGutters,
      $gridUnit: gridUnit,
      $gridGaps: gridGaps
    }, child.props.children);
  }));
}
