// ngast/lib/provider-symbol.js
import { ProviderMeta } from '@angular/compiler';
var ProviderSymbol = /** @class */ (function () {
    function ProviderSymbol(program, provider, metadataResolver) {
        this.program = program;
        this.provider = provider;
        this.metadataResolver = metadataResolver;
    }
    /**
     * Returns the provider metadata.
     */
    ProviderSymbol.prototype.getMetadata = function () {
        return this.provider;
    };
    /**
     * Returns the list of dependencies for given provider.
     */
    ProviderSymbol.prototype.getDependencies = function () {
        var _this = this;
        return (this.provider.deps || []).map(function (d) {
            var token = d.token;
            if (d.token) {
                if (d.token.identifier) {
                    token = d.token.identifier.reference;
                }
            }
            var meta = new ProviderMeta(token, { useClass: d.value });
            return new ProviderSymbol(_this.program, _this.metadataResolver.getProviderMetadata(meta), _this.metadataResolver);
        });
    };
    return ProviderSymbol;
}());
export { ProviderSymbol };
