// baseui/esm/file-uploader/styled-components.js
function _objectSpread2(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/*
Copyright (c) 2018-2020 Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import { styled } from '../styles/index.js';
export var StyledFileDragAndDrop = styled('div', function (props) {
  return {
    alignItems: 'center',
    backgroundColor: props.$isDragActive ? props.$theme.colors.fileUploaderBackgroundColorActive : props.$theme.colors.fileUploaderBackgroundColor,
    borderColor: props.$isDragActive ? props.$theme.colors.fileUploaderBorderColorActive : props.$theme.colors.fileUploaderBorderColorDefault,
    borderStyle: props.$afterFileDrop ? 'none' : 'dashed',
    borderWidth: props.$theme.sizing.scale0,
    boxSizing: 'border-box',
    display: 'flex',
    flexDirection: 'column',
    outline: props.$isDisabled ? 'none' : null,
    paddingTop: props.$theme.sizing.scale900,
    paddingRight: props.$theme.sizing.scale800,
    paddingBottom: props.$afterFileDrop ? props.$theme.sizing.scale300 : props.$theme.sizing.scale900,
    paddingLeft: props.$theme.sizing.scale800,
    width: '100%'
  };
});
StyledFileDragAndDrop.displayName = "StyledFileDragAndDrop";
export var StyledContentMessage = styled('div', function (props) {
  return _objectSpread2({}, props.$theme.typography.font300, {
    color: props.$afterFileDrop ? props.$theme.colors.fileUploaderMessageColor : null,
    marginBottom: props.$afterFileDrop ? props.$theme.sizing.scale600 : null
  });
});
StyledContentMessage.displayName = "StyledContentMessage";
export var StyledErrorMessage = styled('div', function (props) {
  return _objectSpread2({}, props.$theme.typography.font300, {
    color: props.$theme.colors.negative,
    marginBottom: props.$afterFileDrop ? props.$theme.sizing.scale600 : null
  });
});
StyledErrorMessage.displayName = "StyledErrorMessage";
export var StyledRoot = styled('div', function (props) {
  return _objectSpread2({}, props.$theme.typography.font300, {
    color: props.$theme.colors.fileUploaderMessageColor
  });
});
StyledRoot.displayName = "StyledRoot";
export var StyledHiddenInput = styled('input', {});
StyledHiddenInput.displayName = "StyledHiddenInput";
