// arcadable-shared/out/model/instructions/instructionSet.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
exports.__esModule = true;
exports.InstructionSetPointer = exports.InstructionSet = void 0;
var callStack_1 = require("./../callStack");
var logicElement_1 = require("../logicElement");
var instruction_1 = require("./instruction");
var InstructionSet = /** @class */ (function (_super) {
    __extends(InstructionSet, _super);
    function InstructionSet(ID, size, instructions, async, name, game) {
        var _this = _super.call(this, ID, name, game) || this;
        _this.size = size;
        _this.instructions = instructions;
        _this.async = async;
        return _this;
    }
    InstructionSet.prototype.getExecutables = function () {
        return __awaiter(this, void 0, void 0, function () {
            var allExecutables;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.instructions.reduce(function (acc, instruction) { return __awaiter(_this, void 0, void 0, function () {
                            var accumulative, newExecutable_1, newExecutable_2, newExecutable_3, toAwait, newExecutable;
                            var _this = this;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, acc];
                                    case 1:
                                        accumulative = _a.sent();
                                        switch (this.game.instructions[instruction.ID].instructionType) {
                                            case instruction_1.InstructionType.Wait: {
                                                newExecutable_1 = new callStack_1.Executable(function () { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                                                    return [2 /*return*/, instruction.getExecutables(this.async)];
                                                }); }); }, this.async, instruction.await(), [], null, function () { return _this.game.instructions[instruction.ID].amountValue.getValue(); });
                                                return [2 /*return*/, __spreadArrays(accumulative, [newExecutable_1])];
                                            }
                                            case instruction_1.InstructionType.RunCondition: {
                                                newExecutable_2 = new callStack_1.Executable(function () { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                                                    return [2 /*return*/, instruction.getExecutables(this.async)];
                                                }); }); }, this.async, this.async, [], null, null);
                                                return [2 /*return*/, __spreadArrays(accumulative, [newExecutable_2])];
                                            }
                                            case instruction_1.InstructionType.Tone: {
                                                if (instruction.await()) {
                                                    newExecutable_3 = new callStack_1.Executable(function () { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                                                        return [2 /*return*/, instruction.getExecutables(this.async)];
                                                    }); }); }, this.async, false, [], null, null);
                                                    toAwait = new callStack_1.Executable(function () { return __awaiter(_this, void 0, void 0, function () {
                                                        var _this = this;
                                                        return __generator(this, function (_a) {
                                                            return [2 /*return*/, [new callStack_1.Executable(function () { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                                                                        return [2 /*return*/, []];
                                                                    }); }); }, this.async, false, [], null, null)]];
                                                        });
                                                    }); }, this.async, true, [], null, function () { return _this.game.instructions[instruction.ID].durationValue.getValue(); });
                                                    return [2 /*return*/, __spreadArrays(accumulative, [newExecutable_3, toAwait])];
                                                }
                                                break;
                                            }
                                        }
                                        newExecutable = new callStack_1.Executable(function () { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                                            return [2 /*return*/, instruction.getExecutables(this.async)];
                                        }); }); }, this.async, instruction.await(), [], null, null);
                                        return [2 /*return*/, __spreadArrays(accumulative, [newExecutable])];
                                }
                            });
                        }); }, Promise.resolve([]))];
                    case 1:
                        allExecutables = _a.sent();
                        return [2 /*return*/, this.processAwaiting(allExecutables)];
                }
            });
        });
    };
    InstructionSet.prototype.processAwaiting = function (executables) {
        return __awaiter(this, void 0, void 0, function () {
            var awaitExecutable, processedExecutables, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        awaitExecutable = null;
                        processedExecutables = executables.reduce(function (acc, executable) {
                            if (!!awaitExecutable) {
                                awaitExecutable.awaiting.push(executable);
                                return acc;
                            }
                            else {
                                if (executable.await) {
                                    awaitExecutable = executable;
                                }
                                return __spreadArrays(acc, [executable]);
                            }
                        }, []);
                        if (!!!awaitExecutable) return [3 /*break*/, 2];
                        _a = awaitExecutable;
                        return [4 /*yield*/, this.processAwaiting(awaitExecutable.awaiting)];
                    case 1:
                        _a.awaiting = _b.sent();
                        _b.label = 2;
                    case 2: return [2 /*return*/, processedExecutables];
                }
            });
        });
    };
    return InstructionSet;
}(logicElement_1.LogicElement));
exports.InstructionSet = InstructionSet;
var InstructionSetPointer = /** @class */ (function () {
    function InstructionSetPointer(ID, game) {
        this.ID = ID;
        this.game = game;
    }
    InstructionSetPointer.prototype.getExecutables = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.game.instructionSets[this.ID].getExecutables()];
            });
        });
    };
    return InstructionSetPointer;
}());
exports.InstructionSetPointer = InstructionSetPointer;
