// baseui/esm/themes/index.js
function _objectSpread2(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/*
Copyright (c) 2018-2020 Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import createTheme from './creator.js';
import { LightTheme } from './light-theme/light-theme.js';
import { LightThemeMove } from './move-theme/light-theme-with-move.js';
import lightColorTokens from './light-theme/color-tokens.js';
import { fontTokens } from './shared/typography.js';
import { DarkTheme } from './dark-theme/dark-theme.js';
import { DarkThemeMove } from './move-theme/dark-theme-with-move.js';
import darkColorTokens from './dark-theme/color-tokens.js';

var lightThemePrimitives = _objectSpread2({}, lightColorTokens, {}, fontTokens);

var darkThemePrimitives = _objectSpread2({}, darkColorTokens, {}, fontTokens);

export { createTheme, LightTheme, LightThemeMove, lightThemePrimitives, DarkTheme, DarkThemeMove, darkThemePrimitives, DarkTheme as darkThemeOverrides };
export * from './types';
